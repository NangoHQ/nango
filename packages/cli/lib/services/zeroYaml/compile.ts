import chalk from 'chalk';
import { build } from 'esbuild';
import { glob } from 'glob';
import ts from 'typescript';

import { Err, Ok } from '../../utils/result.js';

import type { Result } from '@nangohq/types';

const tsconfig: ts.CompilerOptions = {
    module: ts.ModuleKind.CommonJS,
    target: ts.ScriptTarget.ESNext,
    strict: true,
    esModuleInterop: true,
    skipLibCheck: true,
    forceConsistentCasingInFileNames: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    allowUnusedLabels: false,
    allowUnreachableCode: false,
    exactOptionalPropertyTypes: true,
    noFallthroughCasesInSwitch: true,
    noImplicitOverride: true,
    noImplicitReturns: true,
    noPropertyAccessFromIndexSignature: true,
    noUncheckedIndexedAccess: true,
    noUnusedLocals: true,
    noUnusedParameters: true,
    declaration: false,
    sourceMap: true,
    composite: false,
    checkJs: false,
    noEmit: true
};

function typeCheck({ entryPoints }: { entryPoints: string[] }): Result<boolean> {
    const program = ts.createProgram({
        rootNames: entryPoints,
        options: {
            ...tsconfig
        }
    });

    const diagnostics = ts.getPreEmitDiagnostics(program);
    if (diagnostics.length <= 0) {
        return Ok(true);
    }

    diagnostics.forEach((diagnostic) => {
        const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
        if (diagnostic.file && diagnostic.start != null) {
            const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
            const fileName = diagnostic.file.fileName;
            console.error(chalk.red('err'), '-', `${chalk.blue(fileName)} ${chalk.yellow(`:${line + 1}:${character + 1}`)}`, `\r\n  ${message}\r\n`);
        } else {
            console.error(message);
        }
    });

    console.error(`Found ${diagnostics.length} error${diagnostics.length > 1 ? 's' : ''}`);
    return Err('errors');
}

export async function compileAll({ fullPath }: { fullPath: string }): Promise<Result<boolean>> {
    try {
        const entryPoints = await glob('**/*.ts', { cwd: fullPath, ignore: ['node_modules/**', 'dist/**', 'build/**'] });

        const typechecked = typeCheck({ entryPoints });
        if (typechecked.isErr()) {
            return typechecked;
        }

        await build({
            entryPoints: entryPoints,
            outdir: 'build',
            bundle: false,
            sourcemap: true,
            format: 'cjs',
            target: 'esnext',
            platform: 'node',
            outbase: '.',
            tsconfigRaw: {
                compilerOptions: {
                    module: 'commonjs',
                    target: 'esnext',
                    strict: true,
                    esModuleInterop: true,
                    skipLibCheck: true,
                    forceConsistentCasingInFileNames: true,
                    moduleResolution: 'node',
                    allowUnusedLabels: false,
                    allowUnreachableCode: false,
                    exactOptionalPropertyTypes: true,
                    noFallthroughCasesInSwitch: true,
                    noImplicitOverride: true,
                    noImplicitReturns: true,
                    noPropertyAccessFromIndexSignature: true,
                    noUncheckedIndexedAccess: true,
                    noUnusedLocals: true,
                    noUnusedParameters: true,
                    declaration: false,
                    sourceMap: true,
                    composite: false,
                    checkJs: false
                }
            },
            logLevel: 'info'
        });
        return Ok(true);
    } catch (err) {
        console.error(err);
        return Err('failed_to_compile');
    }
}
