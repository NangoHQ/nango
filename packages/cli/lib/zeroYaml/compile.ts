import fs from 'node:fs';
import path from 'node:path';

import * as babel from '@babel/core';
import chalk from 'chalk';
import { build } from 'esbuild';
import ora from 'ora';
import ts from 'typescript';

import { generateAdditionalExports } from '../services/model.service.js';
import { Err, Ok } from '../utils/result.js';
import { printDebug } from '../utils.js';
import { BabelError, CompileError, customErrors, importRegex, npmPackageRegex, tsconfig, tsconfigString } from './constants.js';
import { buildDefinitions } from './definitions.js';

import type { BabelErrorType } from './constants.js';
import type { Result } from '@nangohq/types';

/**
 * This function is used to compile the code in the integration.
 *
 * It will:
 * - Typecheck the code
 * - Compile the code to .cjs
 * - Rebuild nango.yaml in memory
 */
export async function compileAll({ fullPath, debug }: { fullPath: string; debug: boolean }): Promise<Result<boolean>> {
    let spinner = ora({ text: 'Typechecking' }).start();

    try {
        // Read the index.ts content
        const indexContentResult = await readIndexContent(fullPath);
        if (indexContentResult.isErr()) {
            spinner.fail();
            return Err('failed_to_read_index_ts');
        }
        const indexContent = indexContentResult.value;

        // Get the entry points
        const entryPoints = getEntryPoints(indexContent);
        if (entryPoints.length === 0) {
            spinner.fail();
            console.error("No entry points found in index.ts (e.g., export * from './syncs/github/fetch.js'). Nothing to compile.");
            return Err('no_file');
        }

        printDebug(`Found ${entryPoints.length} entry points in index.ts: ${entryPoints.join(', ')}`, debug);

        // Typecheck the code
        const typechecked = typeCheck({ fullPath, entryPoints });
        if (typechecked.isErr()) {
            spinner.fail();
            return typechecked;
        }

        spinner.succeed();

        // Build the entry points
        const text = `Building ${entryPoints.length} file(s)`;
        spinner = ora({ text }).start();
        printDebug('Building', debug);
        for (const entryPoint of entryPoints) {
            const entryPointFullPath = path.join(fullPath, entryPoint);
            spinner.text = `${text} - ${entryPoint}`;
            printDebug(`Building ${entryPointFullPath}`, debug);

            const buildRes = await compileOne({ entryPoint: entryPointFullPath, projectRootPath: fullPath });
            if (buildRes.isErr()) {
                spinner.fail(`Failed to build ${entryPoint}`);
                console.error(chalk.red(buildRes.error.message));
                return buildRes;
            }
        }

        spinner.succeed();

        // Build and export the definitions
        spinner = ora({ text: `Exporting definitions` }).start();
        const rebuild = await buildDefinitions({ fullPath, debug });
        if (rebuild.isErr()) {
            spinner.fail(`Failed to compile definitions`);
            console.log(chalk.red(rebuild.error.message));
            return Err(rebuild.error);
        }

        generateAdditionalExports({ parsed: rebuild.value, fullPath, debug });

        spinner.succeed();
    } catch (err) {
        console.error(err);

        spinner.fail();
        return Err('failed_to_compile');
    }

    spinner.succeed('Compiled');
    return Ok(true);
}

/**
 * Reads the content of index.ts in the given fullPath.
 */
export async function readIndexContent(fullPath: string): Promise<Result<string>> {
    const indexTsPath = path.join(fullPath, 'index.ts');
    try {
        const indexContent = await fs.promises.readFile(indexTsPath, 'utf8');
        return Ok(indexContent);
    } catch (err) {
        console.error(`Could not read ${indexTsPath}`, err);
        return Err('failed_to_read_index_ts');
    }
}

/**
 * Extracts the entry points from the index.ts content.
 */
export function getEntryPoints(indexContent: string): string[] {
    const entryPoints: string[] = [];
    let match;
    while ((match = importRegex.exec(indexContent)) !== null) {
        const fp = match.groups?.['path'];
        if (!fp) {
            continue;
        }
        entryPoints.push(fp.endsWith('.js') ? fp : `${fp}.js`);
    }
    return entryPoints;
}

/**
 * Runs the typescript compiler to typecheck the code.
 */
function typeCheck({ fullPath, entryPoints }: { fullPath: string; entryPoints: string[] }): Result<boolean> {
    const program = ts.createProgram({
        rootNames: entryPoints.map((file) => path.join(fullPath, file.replace('.js', '.ts'))),
        options: tsconfig
    });

    const diagnostics = ts.getPreEmitDiagnostics(program);
    if (diagnostics.length <= 0) {
        return Ok(true);
    }

    // On purpose
    console.log('');
    for (const diagnostic of diagnostics) {
        const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
        if (diagnostic.file && diagnostic.start != null) {
            const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
            const fileName = diagnostic.file.fileName.replace(`${fullPath}/`, '');
            console.error(chalk.red('err'), '-', `${chalk.blue(fileName)}${chalk.yellow(`:${line + 1}:${character + 1}`)}`, `\r\n  ${message}\r\n`);
        } else {
            console.error(message);
        }
    }

    console.error(`Found ${diagnostics.length} error${diagnostics.length > 1 ? 's' : ''}`);

    return Err('errors');
}

/**
 * Bundles the entry file using esbuild and returns the bundled code as a string (in memory).
 */
export async function bundleFile({ entryPoint }: { entryPoint: string }): Promise<Result<string>> {
    try {
        const res = await build({
            entryPoints: [entryPoint],
            bundle: true,
            sourcemap: 'inline',
            format: 'cjs',
            target: 'esnext',
            platform: 'node',
            logLevel: 'silent',
            treeShaking: true,
            write: false, // Output in memory
            plugins: [
                {
                    name: 'remove-create-wrappers',
                    setup(build: any) {
                        build.onLoad({ filter: /\.ts$/ }, async (args: any) => {
                            const source = await fs.promises.readFile(args.path, 'utf8');
                            const result = await babel.transformAsync(source, {
                                filename: args.path,
                                plugins: [removeCreateWrappersBabelPlugin],
                                parserOpts: { sourceType: 'module', plugins: ['typescript'] },
                                generatorOpts: { decoratorsBeforeExport: true }
                            });
                            return { contents: result?.code ?? source, loader: 'ts' };
                        });
                    }
                },
                {
                    name: 'external-npm-packages',
                    setup(buildInstance) {
                        buildInstance.onResolve({ filter: npmPackageRegex }, (args) => {
                            if (!args.path.startsWith('.') && !path.isAbsolute(args.path)) {
                                return { path: args.path, external: true };
                            }
                            return null; // let esbuild handle other paths
                        });
                    }
                }
            ],
            tsconfigRaw: {
                compilerOptions: tsconfigString
            }
        });
        if (res.errors.length > 0) {
            return Err('failed_to_build');
        }

        const output = res.outputFiles?.[0]?.text || '';
        return Ok(output);
    } catch (err) {
        // We can throw our own error, so we throw one with a custom code that we get back and re-hydrate a new error
        if (err instanceof Error && err.message.includes('nango_')) {
            const code = /(nango_[a-z_]+)/.exec(err.message)?.[1] as BabelErrorType;
            return Err(new CompileError(code, customErrors[code]));
        }
        return Err(new CompileError('failed_to_build_unknown', err instanceof Error ? err.message : 'unknown_error'));
    }
}

/**
 * We use esbuild to compile the code to .cjs.
 * node.vm only supports CJS and we also bundle all imported files in the same file.
 */
export async function compileOne({ entryPoint, projectRootPath }: { entryPoint: string; projectRootPath: string }): Promise<Result<boolean>> {
    const rel = path.relative(projectRootPath, entryPoint);
    // File are compiled to build/integration-type-script-name.cjs
    // Because it's easier to manipulate the files and it's easier in S3
    const outfile = path.join(projectRootPath, 'build', tsToJsPath(rel));

    // Ensure the output directory exists
    await fs.promises.mkdir(path.dirname(outfile), { recursive: true });

    const bundleResult = await bundleFile({ entryPoint });
    if (bundleResult.isErr()) {
        return Err(bundleResult.error);
    }
    try {
        await fs.promises.writeFile(outfile, bundleResult.value, 'utf8');
    } catch (err) {
        console.error(chalk.red(err));
        return Err('failed_to_write_output');
    }
    return Ok(true);
}

export function tsToJsPath(filePath: string) {
    return filePath.replace(/^\.\//, '').replaceAll('/', '_').replaceAll('.js', '.cjs');
}

/**
 * This plugin is used to remove the create wrappers from the exports.
 *
 * Initially, I didn't want to remove this but because our codebase is ESM and node.vm only compiles CJS
 * it was annoying to have to compile the code twice. And have mixed code when publishing.
 * Since the wrapper is only used to stringly type the exports, we can remove it.
 */
function removeCreateWrappersBabelPlugin({ types: t }: { types: typeof babel.types }): babel.PluginObj<any> {
    const allowedExports = ['createAction', 'createSync', 'createOnEvent'];
    return {
        visitor: {
            ExportDefaultDeclaration(path) {
                if ((path.node as any).__transformedByRemoveCreateWrappers) {
                    return;
                }

                const decl = path.node.declaration;
                let calleeName = null;
                let arg = null;

                // Case 1: export default createAction({...})
                if (t.isCallExpression(decl) && t.isIdentifier(decl.callee) && allowedExports.includes(decl.callee.name)) {
                    let varName = '';
                    calleeName = decl.callee.name;
                    arg = decl.arguments[0];
                    if (!t.isObjectExpression(arg)) {
                        throw new BabelError('nango_invalid_function_param');
                    }

                    if (calleeName === 'createAction') varName = 'action';
                    if (calleeName === 'createSync') varName = 'sync';
                    if (calleeName === 'createOnEvent') varName = 'onEvent';

                    // Inject type property
                    arg.properties = [t.objectProperty(t.identifier('type'), t.stringLiteral(varName)), ...arg.properties];
                    const newValue = arg;
                    // Insert: export const <varName> = <newValue>;
                    const exportConst = t.exportNamedDeclaration(t.variableDeclaration('const', [t.variableDeclarator(t.identifier(varName), newValue)]), []);
                    // Insert: export default <varName>;
                    const exportDefault = t.exportDefaultDeclaration(t.identifier(varName));
                    (exportConst as any).__transformedByRemoveCreateWrappers = true;
                    (exportDefault as any).__transformedByRemoveCreateWrappers = true;
                    path.replaceWithMultiple([exportConst, exportDefault]);
                }
                // Case 2: export default action; (or sync/onEvent)
                else if (t.isIdentifier(decl)) {
                    const binding = path.scope.getBinding(decl.name);
                    if (!binding || !binding.path.isVariableDeclarator()) {
                        throw new BabelError('nango_invalid_default_export');
                    }
                    const init = binding.path.node.init;
                    if (!t.isCallExpression(init) || !t.isIdentifier(init.callee) || !allowedExports.includes(init.callee.name)) {
                        throw new BabelError('nango_invalid_default_export');
                    }

                    let varName = '';
                    calleeName = init.callee.name;
                    arg = init.arguments[0];
                    if (!t.isObjectExpression(arg)) {
                        throw new BabelError('nango_invalid_function_param');
                    }
                    if (calleeName === 'createAction') varName = 'action';
                    if (calleeName === 'createSync') varName = 'sync';
                    if (calleeName === 'createOnEvent') varName = 'onEvent';
                    // Inject type property (mutate the object literal)
                    arg.properties = [t.objectProperty(t.identifier('type'), t.stringLiteral(varName)), ...arg.properties];
                    // Replace the variable's initializer with the object literal
                    binding.path.get('init').replaceWith(arg);
                    return;
                } else {
                    throw new BabelError('nango_unsupported_export');
                }
            }
        }
    };
}
