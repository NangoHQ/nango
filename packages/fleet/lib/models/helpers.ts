import crypto from 'crypto';

export function generateImage(): string {
    const charset = '0123456789abcdef';
    const length = 40;
    const randomBytes = new Uint8Array(length);
    crypto.getRandomValues(randomBytes);

    const commitHash = Array.from(randomBytes)
        .map((byte) => charset[byte % charset.length])
        .join('');
    return `generated/image:${commitHash}`;
}
