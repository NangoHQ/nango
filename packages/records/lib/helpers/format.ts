import md5 from 'md5';
import * as uuid from 'uuid';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc.js';
import type { FormattedRecord, UnencryptedRecordData } from '../types.js';
import { Err, Ok } from '@nangohq/utils';
import type { Result } from '@nangohq/utils';

dayjs.extend(utc);

export const formatRecords = ({
    data,
    connectionId,
    model,
    syncId,
    syncJobId,
    softDelete = false
}: {
    data: UnencryptedRecordData[];
    connectionId: number;
    model: string;
    syncId: string;
    syncJobId: number;
    softDelete?: boolean;
}): Result<FormattedRecord[]> => {
    // hashing unique composite key (connection, model, external_id)
    // to generate stable record ids across script executions
    const stableId = (unencryptedData: UnencryptedRecordData): string => {
        const namespace = uuid.v5(`${connectionId}${model}`, uuid.NIL);
        return uuid.v5(`${connectionId}${model}${unencryptedData.id}`, namespace);
    };
    const formattedRecords: FormattedRecord[] = [];
    const now = new Date();
    for (const datum of data) {
        const data_hash = md5(JSON.stringify(datum));

        if (!datum) {
            break;
        }

        if (!datum['id']) {
            const error = new Error(`Missing id field in record: ${JSON.stringify(datum)}. Model: ${model}`);
            return Err(error);
        }

        const formattedRecord: FormattedRecord = {
            id: stableId(datum),
            json: datum,
            // postgresql does not support null bytes in strings
            external_id: String(datum['id']).replaceAll('\x00', ''),
            data_hash,
            model,
            connection_id: connectionId,
            sync_id: syncId,
            sync_job_id: syncJobId
        };

        if (softDelete) {
            const deletedAt = datum['deletedAt'];
            formattedRecord.updated_at = now;
            formattedRecord.deleted_at = deletedAt ? dayjs(deletedAt as string).toDate() : now;
        } else {
            formattedRecord.deleted_at = null;
        }
        formattedRecords.push(formattedRecord);
    }
    return Ok(formattedRecords);
};
