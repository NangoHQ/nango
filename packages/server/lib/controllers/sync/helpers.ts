export function normalizeSyncParams(syncs: (string | { name: string; variant: string })[]): { syncName: string; syncVariant: string }[] {
    return syncs.map((sync) => {
        if (typeof sync === 'string') {
            if (sync.includes('::')) {
                const [name, variant] = sync.split('::');
                return { syncName: name ?? '', syncVariant: variant ?? '' };
            }
            return { syncName: sync, syncVariant: 'base' };
        }

        return { syncName: sync.name, syncVariant: sync.variant };
    });
}
