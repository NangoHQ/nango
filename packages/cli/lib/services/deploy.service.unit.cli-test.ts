import fs from 'node:fs';
import { describe, expect, it } from 'vitest';
import { parse } from './config.service.js';
import deployService from './deploy.service';
import { copyDirectoryAndContents, fixturesPath, getTestDirectory, removeVersion } from '../tests/helpers.js';
import { compileAllFiles } from './compile.service';

describe('package', () => {
    it('should package correctly', async () => {
        const dir = await getTestDirectory('deploy-nested');

        await copyDirectoryAndContents(`${fixturesPath}/nango-yaml/v2/nested-integrations/hubspot`, `${dir}/hubspot`);
        await copyDirectoryAndContents(`${fixturesPath}/nango-yaml/v2/nested-integrations/github`, `${dir}/github`);
        await fs.promises.copyFile(`${fixturesPath}/nango-yaml/v2/nested-integrations/nango.yaml`, `${dir}/nango.yaml`);

        const success = await compileAllFiles({ fullPath: dir, debug: false });
        expect(success).toBe(true);

        const parsing = parse(dir);
        if (parsing.isErr()) {
            throw parsing.error;
        }

        const res = deployService.package({ parsed: parsing.value.parsed!, debug: false, fullPath: dir });
        expect(removeVersion(JSON.stringify(res!.jsonSchema, null, 2))).toMatchSnapshot();
        expect(
            res?.flowConfigs.map((flow) => {
                return {
                    ...flow,
                    fileBody: {
                        ts: flow.fileBody.ts,
                        // Remove tsup comments
                        js: flow.fileBody.js.replaceAll(/\/\/.*/g, '')
                    }
                };
            })
        ).toMatchSnapshot();
    });

    it('should filter by integrationId when specified', async () => {
        const dir = await getTestDirectory('deploy-nested-filtered');

        await copyDirectoryAndContents(`${fixturesPath}/nango-yaml/v2/nested-integrations/hubspot`, `${dir}/hubspot`);
        await copyDirectoryAndContents(`${fixturesPath}/nango-yaml/v2/nested-integrations/github`, `${dir}/github`);
        await fs.promises.copyFile(`${fixturesPath}/nango-yaml/v2/nested-integrations/nango.yaml`, `${dir}/nango.yaml`);

        const success = await compileAllFiles({ fullPath: dir, debug: false });
        expect(success).toBe(true);

        const parsing = parse(dir);
        if (parsing.isErr()) {
            throw parsing.error;
        }

        // Test with hubspot only provided
        const resHubspot = deployService.package({
            parsed: parsing.value.parsed!,
            debug: false,
            fullPath: dir,
            integrationId: 'hubspot'
        });

        // Verify only hubspot integrations are included
        expect(resHubspot).not.toBeNull();
        expect(resHubspot!.flowConfigs.length).toBeGreaterThan(0);
        expect(resHubspot!.flowConfigs.every((flow) => flow.providerConfigKey === 'hubspot')).toBe(true);

        // Verify no github integrations are included
        expect(resHubspot!.flowConfigs.some((flow) => flow.providerConfigKey === 'github')).toBe(false);

        // Test with github
        const resGithub = deployService.package({
            parsed: parsing.value.parsed!,
            debug: false,
            fullPath: dir,
            integrationId: 'github'
        });

        // Verify only github integrations are included
        expect(resGithub).not.toBeNull();
        expect(resGithub!.flowConfigs.length).toBeGreaterThan(0);
        expect(resGithub!.flowConfigs.every((flow) => flow.providerConfigKey === 'github')).toBe(true);

        // Verify no hubspot integrations are included
        expect(resGithub!.flowConfigs.some((flow) => flow.providerConfigKey === 'hubspot')).toBe(false);

        // Test with no integrationId (should include all)
        const resAll = deployService.package({
            parsed: parsing.value.parsed!,
            debug: false,
            fullPath: dir
        });

        expect(resAll).not.toBeNull();
        expect(resAll!.flowConfigs.length).toBeGreaterThan(0);

        // Verify both hubspot and github integrations are included
        const providerKeys = new Set(resAll!.flowConfigs.map((flow) => flow.providerConfigKey));
        expect(providerKeys.size).toBe(2);
        expect(providerKeys.has('hubspot')).toBe(true);
        expect(providerKeys.has('github')).toBe(true);

        expect(resAll!.flowConfigs.length).toBe(resHubspot!.flowConfigs.length + resGithub!.flowConfigs.length);
    });
});
