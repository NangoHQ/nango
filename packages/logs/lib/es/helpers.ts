import { isTest } from '@nangohq/utils';
import { envs } from '../env.js';
import { logger } from '../utils.js';
import { client } from './client.js';
import { getDailyIndexPipeline, indexMessages, policyRetention } from './schema.js';
import { createMessage } from '../models/messages.js';
import { getFormattedMessage } from '../models/helpers.js';

export async function start() {
    if (!envs.NANGO_LOGS_ENABLED) {
        logger.warning('Elasticsearch is disabled, skipping');
        return;
    }

    logger.info('ðŸ”„ Elasticsearch service starting...');

    await migrateMapping();

    logger.info('âœ… Elasticsearch');
}

export async function migrateMapping() {
    try {
        const index = indexMessages;

        logger.info(`Migrating index "${index.index}"...`);

        // -- Policy
        logger.info(`  Updating policy`);
        await client.ilm.putLifecycle(policyRetention());

        // -- Index
        const existsTemplate = await client.indices.existsIndexTemplate({ name: `${index.index}-template` });
        logger.info(`  ${existsTemplate ? 'updating' : 'creating'} index template "${index.index}"...`);

        await client.indices.putIndexTemplate({
            name: `${index.index}-template`,
            index_patterns: `${index.index}.*`,
            template: {
                settings: index.settings!,
                mappings: index.mappings!,
                aliases: { [index.index]: {} }
            }
        });

        // -- Pipeline
        // Pipeline will automatically create an index based on a field
        // In our case we create a daily index based on "createdAt"
        logger.info(`  Updating pipeline`);
        await client.ingest.putPipeline(getDailyIndexPipeline(index.index));

        const existsAlias = await client.indices.exists({ index: index.index });
        if (!existsAlias) {
            // insert a dummy record to create first index
            logger.info(`  Inserting dummy record`);
            await createMessage(getFormattedMessage({ parentId: '-1' }));
        }
    } catch (err) {
        logger.error(err);
        throw new Error('failed_to_init_elasticsearch');
    }
}

export async function deleteIndex({ prefix }: { prefix: string }) {
    if (!isTest) {
        throw new Error('Trying to delete stuff in prod');
    }

    try {
        const indices = await client.cat.indices({ format: 'json' });
        await Promise.all(
            indices.map(async (index) => {
                if (!index.index?.startsWith(prefix)) {
                    return;
                }

                await client.indices.delete({ index: index.index, ignore_unavailable: true });
            })
        );
    } catch (err) {
        logger.error(err);
        throw new Error('failed_to_deleteIndex');
    }
}
