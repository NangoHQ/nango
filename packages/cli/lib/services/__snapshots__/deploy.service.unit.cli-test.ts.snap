// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`package > should package correctly 1`] = `
"{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "Contact": {
      "type": "object",
      "properties": {
        "id": {
          "type": "string"
        }
      },
      "required": [
        "id"
      ],
      "additionalProperties": false
    },
    "Issue": {
      "type": "object",
      "properties": {
        "id": {
          "type": "string"
        }
      },
      "required": [
        "id"
      ],
      "additionalProperties": false
    }
  },
  "$comment": "This file was generated by Nango (vTest)"
}"
`;

exports[`package > should package correctly 2`] = `
[
  {
    "attributes": {},
    "auto_start": true,
    "endpoints": [
      {
        "method": "GET",
        "path": "/hubspot/contacts",
      },
    ],
    "fileBody": {
      "js": ""use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);


var contacts_exports = {};
__export(contacts_exports, {
  default: () => fetchData
});
module.exports = __toCommonJS(contacts_exports);
async function fetchData(nango) {
  await nango.log("Fetching contacts...");
}
",
      "ts": "import type { NangoSync } from '../../models';

export default async function fetchData(nango: NangoSync): Promise<void> {
    await nango.log('Fetching contacts...');
}
",
    },
    "input": undefined,
    "metadata": {
      "scopes": [],
    },
    "model_schema": [
      {
        "fields": [
          {
            "array": false,
            "name": "id",
            "optional": false,
            "tsType": true,
            "value": "string",
          },
        ],
        "name": "Contact",
      },
    ],
    "models": [
      "Contact",
    ],
    "providerConfigKey": "hubspot",
    "runs": "every day",
    "syncName": "contacts",
    "sync_type": "incremental",
    "track_deletes": false,
    "type": "sync",
    "version": "",
    "webhookSubscriptions": [],
  },
  {
    "endpoints": [
      {
        "method": "POST",
        "path": "/hubspot/contact",
      },
    ],
    "fileBody": {
      "js": ""use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);


var create_contact_exports = {};
__export(create_contact_exports, {
  default: () => runAction
});
module.exports = __toCommonJS(create_contact_exports);
async function runAction(nango) {
  await nango.log("Creating contact...");
}
",
      "ts": "import type { NangoAction } from '../../models';

export default async function runAction(nango: NangoAction): Promise<void> {
    await nango.log('Creating contact...');
}
",
    },
    "input": undefined,
    "metadata": {
      "scopes": [],
    },
    "model_schema": [],
    "models": [],
    "providerConfigKey": "hubspot",
    "runs": null,
    "syncName": "create-contact",
    "track_deletes": false,
    "type": "action",
    "version": "",
  },
  {
    "attributes": {},
    "auto_start": true,
    "endpoints": [
      {
        "method": "GET",
        "path": "/github/issues",
      },
    ],
    "fileBody": {
      "js": ""use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);


var issues_exports = {};
__export(issues_exports, {
  default: () => fetchData
});
module.exports = __toCommonJS(issues_exports);
async function fetchData(nango) {
  await nango.log("Fetching issues...");
}
",
      "ts": "import type { NangoSync } from '../../models';

export default async function fetchData(nango: NangoSync): Promise<void> {
    await nango.log('Fetching issues...');
}
",
    },
    "input": undefined,
    "metadata": {
      "scopes": [],
    },
    "model_schema": [
      {
        "fields": [
          {
            "array": false,
            "name": "id",
            "optional": false,
            "tsType": true,
            "value": "string",
          },
        ],
        "name": "Issue",
      },
    ],
    "models": [
      "Issue",
    ],
    "providerConfigKey": "github",
    "runs": "every day",
    "syncName": "issues",
    "sync_type": "incremental",
    "track_deletes": false,
    "type": "sync",
    "version": "",
    "webhookSubscriptions": [],
  },
  {
    "endpoints": [
      {
        "method": "POST",
        "path": "/github/issue",
      },
    ],
    "fileBody": {
      "js": ""use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);


var create_issue_exports = {};
__export(create_issue_exports, {
  default: () => runAction
});
module.exports = __toCommonJS(create_issue_exports);
async function runAction(nango) {
  await nango.log("Creating issue...");
}
",
      "ts": "import type { NangoAction } from '../../models';

export default async function runAction(nango: NangoAction): Promise<void> {
    await nango.log('Creating issue...');
}
",
    },
    "input": undefined,
    "metadata": {
      "scopes": [],
    },
    "model_schema": [],
    "models": [],
    "providerConfigKey": "github",
    "runs": null,
    "syncName": "create-issue",
    "track_deletes": false,
    "type": "action",
    "version": "",
  },
]
`;
